roslibrust/src/ros1/master_client.rs

```diff
@@ -0,0 +1,360 @@
+
+use log::*;
+
+#[derive(thiserror::Error, Debug)]
+pub enum RosMasterError {
+    #[error("Failed to understand xmlrpc message: {0}")]
+    InvalidXmlRpcMessage(#[from] serde_xmlrpc::Error),
+    #[error("Failed to communicate with server: {0}")]
+    ServerCommunicationFailure(#[from] reqwest::Error),
+    #[error("Ros Master Reported an Internal Error: {0}")]
+    MasterError(String),
+}
+
+/// A client that exposes the API hosted by the [rosmaster](http://wiki.ros.org/ROS/Master_API)
+pub struct MasterClient {
+    client: reqwest::Client,
+    // Address at which the rosmaster should be found
+    master_uri: String,
+    // Address at which this node should be reached
+    client_uri: String,
+    // An id for this node
+    id: String,
+}
+
+// TODO: This is failing to decode system state...
+/// These types are really just aliases to make the getSystemState's return type workable
+// #[derive(serde::Deserialize, Debug)]
+// pub struct ServiceList(Vec<(String, Vec<String>)>);
+// #[derive(serde::Deserialize, Debug)]
+// pub struct SubscriberList(Vec<(String, Vec<String>)>);
+// #[derive(serde::Deserialize, Debug)]
+// pub struct PublisherList(Vec<(String, Vec<String>)>);
+// #[derive(serde::Deserialize, Debug)]
+// pub struct SystemState((PublisherList, SubscriberList, ServiceList));
+
+impl MasterClient {
+    /// Constructs a new client for communicating with a ros master
+    /// - master_uri: Expects a fully resolved uri for the master e.g. "http://localhost:11311"
+    /// - client_uri: Expects a fully resolved name where this client should host its xmlrpc at
+    /// - id: A client_id to use when communicating with the master, expected to be a valid ros name e.g. "/my_node"
+    pub async fn new(
+        master_uri: impl Into<String>,
+        client_uri: impl Into<String>,
+        id: impl Into<String>,
+    ) -> Result<MasterClient, RosMasterError> {
+        // Create a client, but we want to verify a valid connection before handing control back,
+        // so we make an initial request and confirm with works before returning
+        let client = MasterClient {
+            client: reqwest::Client::new(),
+            master_uri: master_uri.into(),
+            client_uri: client_uri.into(),
+            id: id.into(),
+        };
+
+        match client.get_uri().await {
+            Ok(_) => Ok(client),
+            Err(e) => Err(e),
+        }
+    }
+
+    async fn post<T: serde::de::DeserializeOwned + std::fmt::Debug>(
+        &self,
+        request: String,
+    ) -> Result<T, RosMasterError> {
+        trace!("Sending master: {request}");
+        let response = self
+            .client
+            .post(&self.master_uri)
+            .body(request)
+            .send()
+            .await?
+            .text()
+            .await?;
+        trace!("Got response: {response}");
+        let parsed: (i8, String, T) = serde_xmlrpc::response_from_str(&response)?;
```

Comment:
Probably makes sense to destructure this tuple with a named variable for each? I don't know what these are by reading it here.